File: silverlight-ports/vector-defence/original-csharp-source/Towers/Tower.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using VectorDefenceSL.Monsters;

namespace VectorDefenceSL.Towers
{
    public abstract class Tower : BoardElement
    {
        public const double TowerRadius = 10;
        public const int MaxLevel = 6;

        protected const double ToolbarAnimationRotationStep = 0.052359877559829883; // 3 degrees

        private bool paused;
        private Stopwatch stopwatch;

        protected Tower(Button btn, TowerBuildingMode mode)
        {
            btn.Tag = mode;
            btn.Content = Container = new Canvas();
            X = Container.ActualWidth / 2;
            Y = Container.ActualHeight / 2;

            btn.MouseEnter += delegate
            {
                Game.TowerAnimation += AnimateToolboxRendering;
            };
            btn.MouseLeave += delegate
            {
                Game.TowerAnimation -= AnimateToolboxRendering;
            };
            PerformRender();
        }

        protected Tower(double x, double y)
        {
            Container = Game.Canvas;
            X = x;
            Y = y;         
            Game.Towers.Add(this);
            Game.Withdraw(Cost);
            Game.TowerAnimation += Animate;
            Game.Paused += Game_Paused;
            Game.Resumed += Game_Resumed;
        }

        public int Level { get; private set; }

        public int Cost { get; protected set; }

        public double Range { get; protected set; }

        protected double MillisecondsBetweenFiring { private get; set; }

        protected bool CanFireNow
        {
            get
            {
                bool canFire = stopwatch == null || stopwatch.ElapsedMilliseconds >= MillisecondsBetweenFiring;
                return canFire;
            }
        }

        protected Canvas Container { get; private set; }

        public abstract void Render();

        public void Upgrade()
        {
            if (Level < MaxLevel)
            {
                Level++;
                UpgradeWork(Level);
                Cost += 50;
            }
        }

        public virtual void RemoveTowerFromCanvas()
        {
            Game.TowerAnimation -= Animate;
            Game.Paused -= Game_Paused;
            Game.Resumed -= Game_Resumed;
        }

        protected static Point CalculateWhereToShoot(Point source, Sprite enemy, double bulletSpeed)
        {
            // target is the inital enemy position, relative to cannon.
            var target = new Point(enemy.X - source.X, enemy.Y - source.Y);
            // calculate 'a' term.
            var a = (bulletSpeed * bulletSpeed) - ((enemy.DX * enemy.DX) + (enemy.DY * enemy.DY));
            // calculate 'b' term.
            var b = (target.X * enemy.DX) + (target.Y * enemy.DY);
            // calculate 'c' term.
            var c = (target.X * target.X) + (target.Y * target.Y);
            // calculate 'd' term.
            var d = (b * b) + (a * c);
            var t = 0d;
            if (d >= 0)
            {
                // calculate the time of impact
                t = (b + Math.Sqrt(d)) / a;
                // negative values are not accepted
                if (t < 0)
                {
                    t = 0;
                }
            }

            // calculate the bullet's target position
            var targetX = enemy.X + (enemy.DX * t);
            var targetY = enemy.Y + (enemy.DY * t);
            return new Point(targetX, targetY);
        }

        protected abstract void Animate();

        protected void TimeFiring()
        {
            stopwatch = Stopwatch.StartNew();
        }

        protected virtual void AnimateToolboxRendering()
        {
        }

        /* from http://www.phaedy.com/gamez/2010/aim-and-hit-with-projectile/
         */

        protected Monster GetClosestMonsterInRange()
        {
            double smallestDistance = double.MaxValue;
            Monster closestMonster = null;
            double fireRangeSquared = Range * Range;

            foreach (var m in Game.CurrentFrameMonsters)
            {
                if (m.MarkedForRemoval)
                {
                    continue;
                }

                var distSquared = Util.GetDistanceSquared(X, Y, m.X, m.Y, Range, fireRangeSquared);
                if (distSquared == -1)
                {
                    continue;
                }

                if (distSquared < smallestDistance)
                {
                    closestMonster = m;
                    smallestDistance = distSquared;
                }
            }

            return closestMonster;
        }

        protected virtual void UpgradeWork(int newLevel)
        {
            int rangeIncrease = newLevel * 4;
            Range += rangeIncrease;
        }

        protected void HandleTowerSelection(object sender, MouseButtonEventArgs e)
        {
            if (Game.SelectedTower != this)
            {
                Game.SetSelectedTower(this);
            }
            e.Handled = true;
        }

        private void PerformRender()
        {
            Render();
        }

        private void Game_Paused()
        {
            if (stopwatch != null)
            {
                paused = true;
                stopwatch.Stop();
            }
        }

        private void Game_Resumed()
        {
            if (paused)
            {
                paused = false;
                stopwatch.Start();
            }
        }
    }
}

File: silverlight-ports/vector-defence/src/Towers/Tower.ts
import { BoardElement } from "../BoardElement";
import { Game } from "../Game";
import { Point } from "../Point";
import { Sprite } from "../Sprite";
import { Stopwatch } from "../Stopwatch";
import { Monster } from "../Monsters/Monster";
import { Util } from "../Util";

export abstract class Tower extends BoardElement {
  static readonly towerRadius = 10;
  static readonly maxLevel = 6;
  protected static readonly toolbarAnimationRotationStep = 0.052359877559829883;

  private paused = false;
  private stopwatch: Stopwatch | null = null;
  level = 0;
  cost = 0;
  range = 0;
  protected millisecondsBetweenFiring = 0;

  protected constructor(x: number, y: number, addToGame = true) {
    super();
    this.x = x;
    this.y = y;
    if (addToGame) {
      Game.towers.push(this);
      Game.withdraw(this.cost);
      Game.towerAnimation.push(this.animateBound);
      Game.paused.push(this.gamePausedBound);
      Game.resumed.push(this.gameResumedBound);
    }
  }

  protected get canFireNow(): boolean {
    return this.stopwatch === null || this.stopwatch.elapsedMilliseconds >= this.millisecondsBetweenFiring;
  }

  abstract render(context: CanvasRenderingContext2D): void;

  upgrade(): void {
    if (this.level < Tower.maxLevel) {
      this.level += 1;
      this.upgradeWork(this.level);
      this.cost += 50;
    }
  }

  removeTowerFromCanvas(): void {
    Game.towerAnimation = Game.towerAnimation.filter((handler) => handler !== this.animateBound);
    Game.paused = Game.paused.filter((handler) => handler !== this.gamePausedBound);
    Game.resumed = Game.resumed.filter((handler) => handler !== this.gameResumedBound);
  }

  containsPoint(x: number, y: number): boolean {
    return Util.isWithinDistance(this.x, this.y, x, y, Tower.towerRadius + 4);
  }

  protected static calculateWhereToShoot(source: Point, enemy: Sprite, bulletSpeed: number): Point {
    const target = new Point(enemy.x - source.x, enemy.y - source.y);
    const a = (bulletSpeed * bulletSpeed) - ((enemy.dx * enemy.dx) + (enemy.dy * enemy.dy));
    const b = (target.x * enemy.dx) + (target.y * enemy.dy);
    const c = (target.x * target.x) + (target.y * target.y);
    const d = (b * b) + (a * c);
    let t = 0;
    if (d >= 0 && a !== 0) {
      t = (b + Math.sqrt(d)) / a;
      if (t < 0) {
        t = 0;
      }
    }
    return new Point(enemy.x + (enemy.dx * t), enemy.y + (enemy.dy * t));
  }

  protected abstract animate(): void;

  protected timeFiring(): void {
    this.stopwatch = Stopwatch.startNew();
  }

  protected animateToolboxRendering(): void {}

  protected getClosestMonsterInRange(): Monster | null {
    let smallestDistance = Number.MAX_VALUE;
    let closestMonster: Monster | null = null;
    const fireRangeSquared = this.range * this.range;

    for (const monster of Game.currentFrameMonsters) {
      if (monster.markedForRemoval) {
        continue;
      }
      const distSquared = Util.getDistanceSquared(this.x, this.y, monster.x, monster.y, this.range, fireRangeSquared);
      if (distSquared === -1) {
        continue;
      }
      if (distSquared < smallestDistance) {
        closestMonster = monster;
        smallestDistance = distSquared;
      }
    }
    return closestMonster;
  }

  protected upgradeWork(newLevel: number): void {
    this.range += newLevel * 4;
  }

  protected drawBaseCircle(context: CanvasRenderingContext2D, fillOuter: string, fillInner = "#000"): void {
    const gradient = context.createRadialGradient(this.x, this.y, 1, this.x, this.y, Tower.towerRadius);
    gradient.addColorStop(0, fillOuter);
    gradient.addColorStop(1, fillInner);
    context.fillStyle = gradient;
    context.strokeStyle = "#fff";
    context.lineWidth = 1.5;
    context.beginPath();
    context.arc(this.x, this.y, Tower.towerRadius, 0, Math.PI * 2);
    context.fill();
    context.stroke();
  }

  private readonly animateBound = (): void => this.animate();
  private readonly gamePausedBound = (): void => this.gamePaused();
  private readonly gameResumedBound = (): void => this.gameResumed();

  private gamePaused(): void {
    if (this.stopwatch !== null) {
      this.paused = true;
      this.stopwatch.stop();
    }
  }

  private gameResumed(): void {
    if (this.paused && this.stopwatch !== null) {
      this.paused = false;
      this.stopwatch.start();
    }
  }
}
